contracts/crowdfund/Crowdfund.sol

```diff
@@ -521,68 +549,71 @@ abstract contract Crowdfund is Implementation, ERC721Receiver, CrowdfundNFT {
         uint96 amount,
         address delegate,
         uint96 previousTotalContributions,
-        bytes memory gateData
+        bytes memory gateData,
+        bool skipDelegation
     ) private {
-        // Require a non-null delegate.
-        if (delegate == address(0)) {
-            revert InvalidDelegateError();
-        }
-        // Must not be blocked by gatekeeper.
-        IGateKeeper _gateKeeper = gateKeeper;
-        if (_gateKeeper != IGateKeeper(address(0))) {
-            if (!_gateKeeper.isAllowed(contributor, gateKeeperId, gateData)) {
-                revert NotAllowedByGateKeeperError(
-                    contributor,
-                    _gateKeeper,
-                    gateKeeperId,
-                    gateData
-                );
+        if (!skipDelegation) {
+            // Require a non-null delegate.
+            if (delegate == address(0)) {
+                revert InvalidDelegateError();
             }
+
+            // Update delegate.
+            // OK if this happens out of cycle.
+            delegationsByContributor[contributor] = delegate;
         }
 
-        // Update delegate.
-        // OK if this happens out of cycle.
-        delegationsByContributor[contributor] = delegate;
         emit Contributed(contributor, amount, delegate, previousTotalContributions);
 
         // OK to contribute with zero just to update delegate.
-        if (amount != 0) {
-            // Only allow contributions while the crowdfund is active.
-            {
-                CrowdfundLifecycle lc = getCrowdfundLifecycle();
-                if (lc != CrowdfundLifecycle.Active) {
-                    revert WrongLifecycleError(lc);
+        if (amount == 0) return;
+
+        // Must not be blocked by gatekeeper.
+        {
+            IGateKeeper _gateKeeper = gateKeeper;
+            if (_gateKeeper != IGateKeeper(address(0))) {
+                if (!_gateKeeper.isAllowed(msg.sender, gateKeeperId, gateData)) {
```

Comment:
It feels pretty bold/risky to make `msg.sender` assumptions in a function this far down :thinking:

Reply:
Maybe add `gateCheck` modifier we can throw on `contribute()` and `contributeFor()` that does this?